addons/account/models/account_code_mapping.py

```diff
@@ -30,30 +37,34 @@ def browse(self, ids=()):
         return super().browse(ids)
 
     def _search(self, domain, offset=0, limit=None, order=None) -> Query:
+        # This method will populate this model's records in cache when the `code_mapping_ids`
+        # One2many on `account_account` is accessed. Any existing records that correspond to the
+        # search domain will be returned, and additional ones will be created in cache as needed.
         match domain:
-            case [('id', 'in', ids)]:
-                return self.browse(sorted(ids))._as_query()
             case [('account_id', 'in', account_ids)]:
-                self_ids = {
-                    f'{account_id},{company.id}'
-                    for company in self.env.companies
+                companies = self.env.user.company_ids
+                existing_code_mappings = self.env.cache.get_records(self, self._fields['account_id']).filtered(
+                    lambda m: m.account_id.id in account_ids and m.company_id in companies
+                )
+                keys = {(m.account_id.id, m.company_id.id) for m in existing_code_mappings}
+                missing_keys = [
+                    (account_id, company_id)
                     for account_id in account_ids
-                }
-                return self.browse(sorted(self_ids))._as_query()
+                    for company_id in companies.ids
+                    if (account_id, company_id) not in keys
+                ]
+                max_existing_id = max(existing_code_mappings.ids, default=0)
+                new_code_mappings = self.browse(range(max_existing_id + 1, max_existing_id + len(missing_keys) + 1))
+
+                self.env.cache.update(new_code_mappings, self._fields['account_id'], [account_id for account_id, _ in missing_keys])
+                self.env.cache.update(new_code_mappings, self._fields['company_id'], [company_id for _, company_id in missing_keys])
+
+                mappings = existing_code_mappings | new_code_mappings
+                return mappings.sorted(lambda m: (m.account_id, m.company_id.sequence, m.company_id.name))._as_query()
```

Comment:
The first element of the tuple doesn't do what you'd expect because `<` on recordsets does set comparison and not value comparison

Reply:
Ah yes, thank you so much! :laughing: 
I'm now just sorting by `account_id.id`, that should be enough to ensure that code mappings of the same account are grouped together.